File: apps/ios/Sources/TenantApp/AppState.swift
import Foundation
import SwiftUI
import TenantKit

/// Persisted, non-sensitive app preferences (the site this app is connected to
/// and the customer's language choice). Tokens never live here — they are in the
/// Keychain via `CustomerSession`.
///
/// A reference type so `AppState` can mutate it without being `var` itself.
final class AppPreferences {
	private let defaults: UserDefaults

	init(defaults: UserDefaults = .standard) {
		self.defaults = defaults
	}

	var siteAddress: SiteAddress? {
		get {
			let slug = defaults.string(forKey: "epic.siteSlug") ?? ""
			let host = defaults.string(forKey: "epic.siteHost") ?? ""
			let origin = defaults.string(forKey: "epic.siteOrigin") ?? ""
			guard !slug.isEmpty || !host.isEmpty else { return nil }
			return SiteAddress(
				slug: slug.isEmpty ? nil : slug,
				host: host.isEmpty ? nil : host,
				origin: URL(string: origin)
			)
		}
		set {
			defaults.set(newValue?.slug ?? "", forKey: "epic.siteSlug")
			defaults.set(newValue?.host ?? "", forKey: "epic.siteHost")
			defaults.set(newValue?.origin?.absoluteString ?? "", forKey: "epic.siteOrigin")
		}
	}

	var languageOverride: String? {
		get {
			let value = defaults.string(forKey: "epic.language") ?? ""
			return value.isEmpty ? nil : value
		}
		set { defaults.set(newValue ?? "", forKey: "epic.language") }
	}
}

/// The app's single source of truth: tenant branding, the customer session, and
/// the profile shown behind login.
@MainActor
final class AppState: ObservableObject {
	enum Phase: Equatable {
		case loading
		/// No tenant bound yet — the customer connects their site.
		case needsSite
		case ready
		case failed(String)
	}

	@Published private(set) var phase: Phase = .loading
	@Published private(set) var organization: PublicOrganization?
	@Published private(set) var theme: SiteTheme = .fallback
	@Published private(set) var profile: CustomerProfile?
	@Published private(set) var isSignedIn = false
	@Published private(set) var needsName = false
	@Published private(set) var isBusy = false
	@Published private(set) var isSavingProfile = false
	@Published private(set) var didSaveProfile = false
	@Published private(set) var errorMessage: String?
	@Published var pendingPhone: String?

	private(set) var configuration: TenantConfiguration
	private var client: TenantAPIClient
	private var session: CustomerSession?
	private let preferences: AppPreferences

	init(
		configuration: TenantConfiguration = .from(infoDictionary: Bundle.main.infoDictionary),
		preferences: AppPreferences = AppPreferences()
	) {
		self.configuration = configuration
		self.preferences = preferences

		// A white-label build is bound to one tenant at build time; that binding
		// wins over anything a customer typed into an earlier un-branded install,
		// so the app never asks (or re-points) in the shipping case.
		let buildAddress = configuration.siteAddress
		self.configuration.siteAddress = SiteAddress.resolve(
			build: buildAddress,
			stored: preferences.siteAddress
		)
		if buildAddress.isBound {
			preferences.siteAddress = nil
		}

		self.client = TenantAPIClient(configuration: self.configuration)
	}

	// MARK: - Branding

	var siteName: String { organization?.name ?? "Tenant" }

	var iconURL: URL? { organization?.iconURL(baseURL: configuration.appBaseURL) }

	var announcements: [PublicSiteAnnouncement] { organization?.announcements ?? [] }

	/// UI language: customer override → device language → org language → English.
	var language: AppLanguage {
		AppLanguage.resolve(
			organization: organization,
			deviceLanguages: Locale.preferredLanguages,
			appLocales: AppLocalizations.available,
			override: preferences.languageOverride
		)
	}

	var languageOverride: String? {
		get { preferences.languageOverride }
		set {
			preferences.languageOverride = newValue
			objectWillChange.send()
		}
	}

	var availableLanguages: [String] { AppLocalizations.available }

	// MARK: - Lifecycle

	func start() async {
		guard configuration.isBoundToTenant else {
			phase = .needsSite
			return
		}
		await loadSite()
	}

	/// Connects the app to a tenant the customer typed in.
	///
	/// The binding is only persisted once the tenant's payload resolves, so a
	/// typo cannot leave the app pointed at a site that does not exist.
	func connect(to input: String) async {
		guard let address = SiteAddress.parse(input, brandDomain: configuration.brandDomain) else {
			errorMessage = language.string("connect.error")
			return
		}
		await loadSite(bindingTo: address)
	}

	func retry() async {
		guard configuration.isBoundToTenant else {
			phase = .needsSite
			return
		}
		await loadSite()
	}

	private func loadSite(bindingTo address: SiteAddress? = nil) async {
		let previousAddress = configuration.siteAddress
		if let address {
			// The client captures the configuration, so rebuild it for the new binding.
			configuration.siteAddress = address
			client = TenantAPIClient(configuration: configuration)
		}

		phase = .loading
		errorMessage = nil
		do {
			let organization = try await client.fetchOrganization()
			if let address {
				preferences.siteAddress = address
			}
			self.organization = organization
			self.theme = SiteTheme(theme: organization.theme)
			configureSession(for: organization)
			phase = .ready
		} catch {
			if address != nil {
				// Roll back for any failure (decoding errors included) so a failed
				// attempt cannot strand the app on a binding that does not resolve.
				configuration.siteAddress = previousAddress
				client = TenantAPIClient(configuration: configuration)
				phase = previousAddress.isBound ? .failed(message(for: error)) : .needsSite
			} else {
				phase = .failed(message(for: error))
			}
		}
	}

	/// Rebuilds the API client for the org's data region and restores any session.
	private func configureSession(for organization: PublicOrganization) {
		configuration.dataRegion = organization.resolvedDataRegion
		let client = TenantAPIClient(configuration: configuration)
		let session = CustomerSession(client: client)
		self.client = client
		self.session = session

		Task {
			let tokens = await session.restore()
			// A stored session belongs to one org; drop it when the tenant changes.
			if let tokens, tokens.orgId != organization.id {
				await session.signOut()
				isSignedIn = false
				return
			}
			isSignedIn = tokens != nil
			if tokens != nil { await loadProfile() }
		}
	}

	// MARK: - Authentication

	func sendCode(phone: String) async {
		guard let session else { return }
		guard PhoneNumber.isValid(phone) else {
			errorMessage = language.string("login.invalidPhone")
			return
		}
		isBusy = true
		errorMessage = nil
		do {
			try await session.sendCode(phone: phone)
			pendingPhone = PhoneNumber.normalize(phone)
		} catch let error as APIError {
			errorMessage = message(for: error)
		} catch {
			errorMessage = language.string("login.networkError")
		}
		isBusy = false
	}

	func verify(code: String) async {
		guard let session, let phone = pendingPhone else { return }
		isBusy = true
		errorMessage = nil
		do {
			let result = try await session.verify(phone: phone, code: code)
			isSignedIn = true
			if result.needsName == true {
				needsName = true
			} else {
				await loadProfile()
			}
		} catch let error as APIError {
			errorMessage = message(for: error)
		} catch {
			errorMessage = language.string("login.networkError")
		}
		isBusy = false
	}

	func resendCode() async {
		guard let phone = pendingPhone else { return }
		await sendCode(phone: phone)
	}

	func cancelVerification() {
		pendingPhone = nil
		errorMessage = nil
	}

	func completeName(_ name: String) async {
		guard let session else { return }
		isBusy = true
		errorMessage = nil
		do {
			profile = try await session.updateProfile(name: name, email: nil)
			needsName = false
		} catch let error as APIError {
			errorMessage = message(for: error)
		} catch {
			errorMessage = language.string("login.networkError")
		}
		isBusy = false
	}

	func signOut() async {
		await session?.signOut()
		isSignedIn = false
		profile = nil
		needsName = false
		pendingPhone = nil
		errorMessage = nil
	}

	// MARK: - Profile

	func loadProfile() async {
		guard let session, isSignedIn else { return }
		do {
			let profile = try await session.profile()
			self.profile = profile
			if profile.needsName == true { needsName = true }
		} catch let error as APIError where error.isUnauthorized {
			isSignedIn = false
			profile = nil
		} catch {
			errorMessage = message(for: error)
		}
	}

	func saveProfile(name: String, email: String) async {
		guard let session else { return }
		isSavingProfile = true
		errorMessage = nil
		didSaveProfile = false
		do {
			profile = try await session.updateProfile(name: name, email: email)
			didSaveProfile = true
		} catch let error as APIError {
			errorMessage = message(for: error)
		} catch {
			errorMessage = language.string("profile.networkError")
		}
		isSavingProfile = false
	}

	func clearTransientMessages() {
		errorMessage = nil
		didSaveProfile = false
	}

	// MARK: - Helpers

	/// Server messages are English; transport failures get a localized one.
	private func message(for error: Error) -> String {
		guard let apiError = error as? APIError else {
			return language.string("login.networkError")
		}
		if apiError.isTransportFailure {
			return language.string("login.networkError")
		}
		return apiError.message
	}
}

File: apps/android/app/src/main/kotlin/com/epicstartup/tenant/ui/AppState.kt
package com.epicstartup.tenant.ui

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Configuration
import com.epicstartup.tenant.R
import com.epicstartup.tenant.core.config.TenantConfiguration
import com.epicstartup.tenant.core.localization.AppLanguage
import com.epicstartup.tenant.core.localization.SiteLocale
import com.epicstartup.tenant.core.model.CustomerProfile
import com.epicstartup.tenant.core.model.PublicOrganization
import com.epicstartup.tenant.core.model.PublicSiteAnnouncement
import com.epicstartup.tenant.core.net.ApiException
import com.epicstartup.tenant.core.net.TenantApiClient
import com.epicstartup.tenant.core.session.CustomerSession
import com.epicstartup.tenant.core.support.PhoneNumber
import com.epicstartup.tenant.core.support.SiteAddress
import com.epicstartup.tenant.core.theme.SiteTheme
import com.epicstartup.tenant.platform.AndroidTenantConfiguration
import com.epicstartup.tenant.platform.AppPreferences
import com.epicstartup.tenant.platform.KeystoreTokenStorage
import com.epicstartup.tenant.platform.TaskRunner
import java.util.Locale
import java.util.concurrent.CopyOnWriteArrayList

/**
 * The app's single source of truth: tenant branding, the customer session, and
 * the profile shown behind login (mirrors `AppState` in apps/ios).
 *
 * One instance lives for the process (see [AppStateHolder]) so an Activity
 * recreation — a language switch, a rotation, a process restore — never
 * refetches branding or drops a half-typed form. Mutations happen on the main
 * thread: callers enter from the UI, and every background result is posted back
 * by [TaskRunner].
 */
class AppState(private val context: Context) {

	sealed interface Phase {
		data object Loading : Phase
		/** No tenant bound yet — the customer connects their site. */
		data object NeedsSite : Phase
		data object Ready : Phase
		data class Failed(val message: String) : Phase
	}

	fun interface Listener {
		fun onStateChanged()
	}

	private val listeners = CopyOnWriteArrayList<Listener>()
	private val preferences = AppPreferences(context)

	var phase: Phase = Phase.Loading
		private set
	var organization: PublicOrganization? = null
		private set
	var theme: SiteTheme = SiteTheme.fallback
		private set
	var profile: CustomerProfile? = null
		private set
	var isSignedIn: Boolean = false
		private set
	var needsName: Boolean = false
		private set
	var isBusy: Boolean = false
		private set
	var isSavingProfile: Boolean = false
		private set
	var didSaveProfile: Boolean = false
		private set
	var errorMessage: String? = null
		private set
	var pendingPhone: String? = null
		private set

	private var configuration: TenantConfiguration = AndroidTenantConfiguration.from(context)
	private var client = TenantApiClient(configuration)
	private var session: CustomerSession? = null
	private var started = false

	/**
	 * Bumped whenever the sign-in attempt is abandoned, so a late OTP response
	 * cannot sign the customer in after they backed out.
	 */
	private var authAttempt = 0

	private var localizedCache: Pair<String, Context>? = null

	init {
		// A white-label build is bound to one tenant at build time; that binding
		// wins over anything a customer typed into an earlier un-branded install,
		// so the app never asks (or re-points) in the shipping case.
		val buildAddress = configuration.siteAddress
		configuration = configuration.withSiteAddress(
			SiteAddress.resolve(build = buildAddress, stored = preferences.siteAddress),
		)
		if (buildAddress.isBound) preferences.siteAddress = null
		client = TenantApiClient(configuration)
	}

	// MARK: - Branding

	val siteName: String get() = organization?.name ?: "Tenant"

	val iconUrl: String? get() = organization?.iconUrl(configuration.appBaseUrl)

	val announcements: List<PublicSiteAnnouncement> get() = organization?.announcements.orEmpty()

	val siteAddress: SiteAddress get() = configuration.siteAddress

	/** UI language: customer override → device language → org language → English. */
	val language: AppLanguage
		get() = AppLanguage.resolve(
			organization = organization,
			deviceLanguages = deviceLanguages(),
			appLocales = SiteLocale.contentLocales,
			override = preferences.languageOverride,
		)

	var languageOverride: String?
		get() = preferences.languageOverride
		set(value) {
			if (preferences.languageOverride == value) return
			preferences.languageOverride = value
			notifyChanged()
		}

	val availableLanguages: List<String> get() = SiteLocale.contentLocales

	// MARK: - Lifecycle

	fun start() {
		if (started) return
		started = true
		if (!configuration.isBoundToTenant) {
			phase = Phase.NeedsSite
			notifyChanged()
			return
		}
		loadSite()
	}

	/** Connects the app to a tenant the customer typed in. */
	fun connect(input: String) {
		val address = SiteAddress.parse(input, configuration.brandDomain)
		if (address == null) {
			errorMessage = string(R.string.connect_error)
			notifyChanged()
			return
		}
		loadSite(bindingTo = address)
	}

	fun retry() {
		if (!configuration.isBoundToTenant) {
			phase = Phase.NeedsSite
			notifyChanged()
			return
		}
		loadSite()
	}

	/**
	 * Loads the org's published branding.
	 *
	 * The binding is only persisted once the tenant's payload resolves, so a
	 * typo cannot leave the app pointed at a site that does not exist.
	 */
	private fun loadSite(bindingTo: SiteAddress? = null) {
		val previousAddress = configuration.siteAddress
		if (bindingTo != null) {
			configuration = configuration.withSiteAddress(bindingTo)
			client = TenantApiClient(configuration)
		}

		phase = Phase.Loading
		errorMessage = null
		notifyChanged()

		val targetClient = client
		TaskRunner.run(
			work = {
				targetClient.fetchOrganization(
					locale = null,
					acceptLanguage = deviceLanguages().joinToString(", "),
				)
			},
			onSuccess = { organization ->
				if (bindingTo != null) preferences.siteAddress = bindingTo
				this.organization = organization
				this.theme = SiteTheme(organization.theme)
				configureSession(organization)
				// `notifyChanged` below also lets listeners notice that the
				// negotiated language changed (MainActivity recreates itself then).
				phase = Phase.Ready
				notifyChanged()
			},
			onFailure = { error ->
				if (bindingTo != null) {
					// Roll back for any failure (decoding errors included) so a
					// failed attempt cannot strand the app on a binding that does
					// not resolve.
					configuration = configuration.withSiteAddress(previousAddress)
					client = TenantApiClient(configuration)
					phase = if (previousAddress.isBound) {
						Phase.Failed(messageFor(error))
					} else {
						Phase.NeedsSite
					}
					errorMessage = messageFor(error)
				} else {
					phase = Phase.Failed(messageFor(error))
				}
				notifyChanged()
			},
		)
	}

	/** Rebuilds the API client for the org's data region and restores any session. */
	private fun configureSession(organization: PublicOrganization) {
		configuration = configuration.withDataRegion(organization.resolvedDataRegion)
		val client = TenantApiClient(configuration)
		this.client = client

		TaskRunner.run(
			// The session is built here, not on the main thread: its constructor
			// reads SharedPreferences and unwraps the Android Keystore.
			work = {
				val session = CustomerSession(client, KeystoreTokenStorage(context))
				val tokens = session.restore()
				// A stored session belongs to one org; drop it when the tenant
				// changes. It is cleared locally on purpose — the tokens may
				// belong to a tenant in another region, and this org's node must
				// never see them.
				if (tokens != null && tokens.orgId != organization.id) {
					session.clearLocalSession()
					null to session
				} else {
					tokens to session
				}
			},
			onSuccess = { (tokens, session) ->
				this.session = session
				isSignedIn = tokens != null
				notifyChanged()
				if (tokens != null) loadProfile()
			},
		)
	}

	// MARK: - Authentication

	fun sendCode(phone: String) {
		val session = session ?: return
		if (!PhoneNumber.isValid(phone)) {
			errorMessage = string(R.string.login_invalid_phone)
			notifyChanged()
			return
		}
		isBusy = true
		errorMessage = null
		notifyChanged()

		TaskRunner.run(
			work = { session.sendCode(phone) },
			onSuccess = {
				pendingPhone = PhoneNumber.normalize(phone)
				isBusy = false
				notifyChanged()
			},
			onFailure = { error ->
				errorMessage = messageFor(error)
				isBusy = false
				notifyChanged()
			},
		)
	}

	fun verify(code: String) {
		val session = session ?: return
		val phone = pendingPhone ?: return
		authAttempt += 1
		val attempt = authAttempt
		isBusy = true
		errorMessage = null
		notifyChanged()

		TaskRunner.run(
			work = { session.verify(phone = phone, code = code) },
			onSuccess = { result ->
				if (attempt != authAttempt) return@run
				isBusy = false
				isSignedIn = true
				if (result.needsName == true) {
					needsName = true
					notifyChanged()
				} else {
					notifyChanged()
					loadProfile()
				}
			},
			onFailure = { error ->
				if (attempt != authAttempt) return@run
				isBusy = false
				errorMessage = messageFor(error)
				notifyChanged()
			},
		)
	}

	fun resendCode() {
		val phone = pendingPhone ?: return
		sendCode(phone)
	}

	fun cancelVerification() {
		// Abandon the attempt: an in-flight `verify` that lands later must not
		// sign the customer in, and its error must not reappear.
		authAttempt += 1
		pendingPhone = null
		errorMessage = null
		isBusy = false
		notifyChanged()
	}

	fun completeName(name: String) {
		val session = session ?: return
		isBusy = true
		errorMessage = null
		notifyChanged()

		TaskRunner.run(
			work = { session.updateProfile(name = name, email = null) },
			onSuccess = { profile ->
				this.profile = profile
				needsName = false
				isBusy = false
				notifyChanged()
			},
			onFailure = { error ->
				errorMessage = messageFor(error)
				isBusy = false
				notifyChanged()
			},
		)
	}

	fun signOut() {
		authAttempt += 1
		val session = session
		isSignedIn = false
		profile = null
		needsName = false
		pendingPhone = null
		errorMessage = null
		notifyChanged()

		if (session != null) TaskRunner.run(work = { session.signOut() })
	}

	// MARK: - Profile

	fun loadProfile() {
		val session = session ?: return
		if (!isSignedIn) return

		TaskRunner.run(
			work = { session.profile() },
			onSuccess = { profile ->
				this.profile = profile
				if (profile.needsName == true) needsName = true
				notifyChanged()
			},
			onFailure = { error ->
				if (error is ApiException && error.isUnauthorized) {
					isSignedIn = false
					profile = null
				} else {
					errorMessage = messageFor(error)
				}
				notifyChanged()
			},
		)
	}

	fun saveProfile(name: String, email: String) {
		val session = session ?: return
		isSavingProfile = true
		errorMessage = null
		didSaveProfile = false
		notifyChanged()

		TaskRunner.run(
			work = { session.updateProfile(name = name, email = email) },
			onSuccess = { profile ->
				this.profile = profile
				isSavingProfile = false
				didSaveProfile = true
				notifyChanged()
			},
			onFailure = { error ->
				errorMessage = messageFor(error, R.string.profile_network_error)
				isSavingProfile = false
				notifyChanged()
			},
		)
	}

	fun clearTransientMessages() {
		if (errorMessage == null && !didSaveProfile) return
		errorMessage = null
		didSaveProfile = false
		notifyChanged()
	}

	// MARK: - Listeners

	fun addListener(listener: Listener) {
		listeners += listener
	}

	fun removeListener(listener: Listener) {
		listeners -= listener
	}

	private fun notifyChanged() {
		listeners.forEach { it.onStateChanged() }
	}

	// MARK: - Helpers

	/** Localized strings for state messages (the UI has its own themed context). */
	fun string(id: Int): String = localizedContext().getString(id)

	private fun localizedContext(): Context {
		val code = language.code
		localizedCache?.let { if (it.first == code) return it.second }
		val configuration = Configuration(context.resources.configuration)
		val locale = Locale.forLanguageTag(code)
		configuration.setLocale(locale)
		configuration.setLayoutDirection(locale)
		val created = context.createConfigurationContext(configuration)
		localizedCache = code to created
		return created
	}

	private fun deviceLanguages(): List<String> =
		android.os.LocaleList.getDefault().toLanguageTags().split(',').map { it.trim() }

	/** Server messages are English; transport failures get a localized one. */
	private fun messageFor(error: Exception, networkErrorRes: Int = R.string.login_network_error): String {
		val apiError = error as? ApiException ?: return string(networkErrorRes)
		if (apiError.isTransportFailure) return string(networkErrorRes)
		return apiError.message
	}
}

/**
 * Process-wide holder so an Activity recreation keeps the loaded tenant and
 * session (the Android equivalent of iOS keeping `@StateObject` alive).
 */
object AppStateHolder {
	// The singleton deliberately holds the *application* context (never an
	// Activity), so there is nothing to leak across a configuration change.
	@SuppressLint("StaticFieldLeak")
	@Volatile
	private var state: AppState? = null

	fun get(context: Context): AppState =
		state ?: synchronized(this) {
			state ?: AppState(context.applicationContext).also { state = it }
		}
}
